### DISCUSSION TITLE
Struggling to understand how FieldArray V7 works

### DISCUSSION THREAD

**Discussion by alveshelio** [react-hook-form/react-hook-form#5309] [Q&A] [answered]
Hi guys,
 
I'm working on a form with FieldArray and I'm having a hard time understanding how it works.

In my form I have the ability to add items to an array, an item has `{name: string, yearsOfExperience: string }` and if I add one and select an option, then add another and select another option then there's not issue. The problem is when I add two or 3 in a row without selection any option, then if I console.log the items, in the name of the item, which should be a string, it gets assigned an object `{name: string, yearsOfExperience: string }`

As you can see here: 
![image](https://user-images.githubusercontent.com/8176422/119341094-4b9d2d00-bc8b-11eb-9814-7317f55f524e.png)
In the first item, name is a string, however on the second, the name is an object.

I have created this CodeSandbaox https://codesandbox.io/s/competent-pasteur-dl8ht?file=/src/index.tsx

I was following this video: https://www.youtube.com/watch?v=LhIvi4XHIco which is for V6 but I've tried to implement it for version 7.
What confuses me is that on version 6 when you map through `fields` you get `{ id, name, amount, type }` because when you do 
```
const {
    fields,
    append,
    remove,
  } = useFieldArray({
    control,
    name: 'programmingLanguages',
  })
```
its seems that in v6 you get both fields `name` and `yearsOfExperience` however in V7 you only get the `id`
I'm using `useController` to register both the select and the input and looking at what's done in V6 there's no need for it.

I'm wondering if I'm over complicating things or what I'm missing :(

Any help would be greatly appreciated :)

**Comment by bluebill1049** [comment]
> its seems that in v6 you get both fields name and yearsOfExperience however in V7 you only get the id
I'm using useController to register both the select and the input and looking at what's done in V6 there's no need for it.

what's the exact problem here?

**Reply by alveshelio** [reply]
I'm not saying it's a problem :) 
It's just that I'm struggling to understand how to implement fieldArray correctly and in V6 if you had a value for example `programmingLanguage` that would be an object with `name` and `yearsOfExperience` you'd get access to both fields when using `useController`, however now in V7 I only get the id of the field.

And when I append two or more fields then the `name` instead of getting '', it gets assigned with an object `name` and `yearsOfExperience` as you can see in the image above.

Thank you for your help. It's really appreciated.

**Reply by bluebill1049** [reply]
<img width="1092" alt="Screen Shot 2021-05-28 at 7 53 35 pm" src="https://user-images.githubusercontent.com/10513364/119966175-64407680-bfee-11eb-920d-5da475de1a8b.png">

when I run your app, I got the error above. I am too sure why the nested name is happening, feel like something got nested. is it happen during append?

**Reply by alveshelio** [reply]
I've fixed the missing import.
And you are correct, it does happen during append.

**Reply by bluebill1049** [reply]
<img width="505" alt="Screen Shot 2021-05-29 at 8 07 36 pm" src="https://user-images.githubusercontent.com/10513364/120066407-85ba6480-c0b9-11eb-9c03-6da9a16a0360.png">

it looks fine to me in the updated csb?

**Reply by alveshelio** [reply]
When you update the fields with a value it's all good. The problem is if you simply append fields with no values and you hit submit, I'm using yup to validate the form and if I add multiple fields without values and hit submit I get this error message:
`techTools[0].name must be a `string` type, but the final value was: `{ "name": "\"\"", "yearsOfExperience": "null" }`.
You need to add the years of experience`
Because field 'name' is an object instead of a string.

**Comment by leapful** [comment, marked as answer]
@alveshelio 

You made a lot of bugs inside your complex implementation.

1. In ToolSelect.tsx, line 108, the handling "useEffect" always sets "selectName" which is the Autocomplete for field "name" to an object { name, yearsOfExperience }, that's why the yup validation alerts about wrong type for field "name".
2. In Autocomplete.tsx, the "selectName" is not set to attribute "name" of Input component, therefore, react-hook-form cannot get the correct field name to handle.

Working sample: https://codesandbox.io/s/agitated-surf-ftdt8?file=/src/ToolSelect.tsx

P.S: You might want to start with simple input and apply form handling first to make sure all form functionalities work then start implementing extra handling in order to avoid struggling inside your own implementation.

**Reply by alveshelio** [reply]
Thank you very much for the precious help. I'm going to compare your implementation with mine to understand my mistakes and learn from it.

**Reply by bluebill1049** [reply]
Thank you very much @leapful
